# Supervised Learning Homework

## Phase 1: Logistic Regression vs. Random Forest Hypothesis (unscaled data)

The hypothesis is that random forest will be a better model than linear regression.
The random forest is less influenced by outlieirs, and there is no underlying assumption of linearity in the data.


In [98]:
# Import required modules

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler

In [101]:
# Read the 2019 Loans CSV file into a Pandas DataFrame

py_loans_df = pd.read_csv('Resources/2019loans.csv')
py_loans_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0         7000.0    0.1894       256.38       MORTGAGE     75000.0   
1        40000.0    0.1614       975.71       MORTGAGE    102000.0   
2        11000.0    0.2055       294.81           RENT     45000.0   
3         4000.0    0.1612       140.87       MORTGAGE     38000.0   
4        14000.0    0.1797       505.93       MORTGAGE     43000.0   
...          ...       ...          ...            ...         ...   
12175    19975.0    0.2565       801.09           RENT     28000.0   
12176    15000.0    0.1774       540.34           RENT     50000.0   
12177     3600.0    0.1862       131.28           RENT     60000.0   
12178    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0            Not Verified          n  28.62          0.0             2.0  ...   
1         Source Verified          n  11.72          2.0             0.0  ...   
2                Verified          n  37.25          1.0             3.0  ...   
3            Not Verified          n  42.89          1.0             0.0  ...   
4         Source Verified          n  22.16          1.0             0.0  ...   
...                   ...        ...    ...          ...             ...  ...   
12175        Not Verified          n  28.42          0.0             0.0  ...   
12176            Verified          n  23.43          4.0             0.0  ...   
12177        Not Verified          n  28.80          0.0             1.0  ...   
12178     Source Verified          n  11.44          0.0             0.0  ...   
12179        Not Verified          n  21.66          1.0             1.0  ...   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  87.5                   0.0        0.0         352260.0   
1                   0.0                   0.0        0.0         294664.0   
2                   7.7                   0.0        0.0          92228.0   
3                 100.0                   0.0        0.0         284273.0   
4                  25.0                   0.0        0.0         120280.0   
...                 ...                   ...        ...              ...   
12175              16.7                   0.0        0.0          50055.0   
12176              11.1                   0.0        0.0          70324.0   
12177               0.0                   0.0        0.0          83765.0   
12178               0.0                   0.0        0.0         189930.0   
12179             100.0                   0.0        0.0          48968.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0               62666.0         35000.0                     10000.0   
1              109911.0          9000.0                     71044.0   
2               36007.0         33000.0                     46328.0   
3               52236.0         13500.0                     52017.0   
4               88147.0         33300.0                     78680.0   
...                 ...             ...                         ...   
12175           28192.0         18700.0                     19055.0   
12176           57025.0         13300.0                     54824.0   
12177           55156.0         14800.0                     53065.0   
12178           23748.0          7000.0                     32930.0   
12179           29786.0          2900.0                     33968.0   

       hardship_flag  debt_settlement_flag     target  
0                  N                     N   low_risk  
1                  N                     N   low_risk  
2                  N                     N   low_risk  
3                  N                     N   low_risk  
4                  N                     N   low_risk  
...              ...                   ...        ...  
12175              N   

In [102]:
# Drop the target column and create the 2019 fratures DataFrame
X_py = py_loans_df.drop('target', axis=1)
X_py.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     7000.0    0.1894       256.38       MORTGAGE     75000.0   
1    40000.0    0.1614       975.71       MORTGAGE    102000.0   
2    11000.0    0.2055       294.81           RENT     45000.0   
3     4000.0    0.1612       140.87       MORTGAGE     38000.0   
4    14000.0    0.1797       505.93       MORTGAGE     43000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  28.62          0.0             2.0  ...   
1     Source Verified          n  11.72          2.0             0.0  ...   
2            Verified          n  37.25          1.0             3.0  ...   
3        Not Verified          n  42.89          1.0             0.0  ...   
4     Source Verified          n  22.16          1.0             0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0           100.0              87.5                   0.0        0.0   
1            85.7               0.0                   0.0        0.0   
2            94.3               7.7                   0.0        0.0   
3            92.9             100.0                   0.0        0.0   
4            84.1              25.0                   0.0        0.0   

  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0        352260.0            62666.0         35000.0   
1        294664.0           109911.0          9000.0   
2         92228.0            36007.0         33000.0   
3        284273.0            52236.0         13500.0   
4        120280.0            88147.0         33300.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     10000.0              N                     N  
1                     71044.0              N                     N  
2                     46328.0              N                     N  
3                     52017.0              N                     N  
4                     78680.0              N                     N  

[5 rows x 83 columns]

In [103]:
# Read the 2020 Q1 loans CSV into a DataFrame
q1_loans_df = pd.read_csv('Resources/2020Q1loans.csv')
q1_loans_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0       40000.0    0.1033       856.40           RENT    128700.0   
1       24450.0    0.1430       572.72       MORTGAGE     44574.0   
2       13500.0    0.1430       316.23            OWN     60000.0   
3       10625.0    0.1774       268.31           RENT     60000.0   
4        6375.0    0.1862       232.46           RENT     60000.0   
...         ...       ...          ...            ...         ...   
4697    30000.0    0.1240       673.42           RENT    140480.0   
4698    24000.0    0.0756       747.22           RENT     50000.0   
4699    10000.0    0.2305       387.36           RENT     33000.0   
4700     8000.0    0.1862       205.86           RENT     38000.0   
4701    30000.0    0.2055      1123.34           RENT    180000.0   

     verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Source Verified          n  12.47          0.0             1.0  ...   
1           Not Verified          n  15.05          0.0             1.0  ...   
2           Not Verified          n  28.72          0.0             0.0  ...   
3               Verified          n  15.70          0.0             4.0  ...   
4        Source Verified          n  35.50          0.0             0.0  ...   
...                  ...        ...    ...          ...             ...  ...   
4697     Source Verified          n  15.74          0.0             0.0  ...   
4698        Not Verified          n  26.81          0.0             0.0  ...   
4699            Verified          n  38.51          0.0             2.0  ...   
4700     Source Verified          n  16.36          0.0             1.0  ...   
4701     Source Verified          n  12.06          0.0             0.0  ...   

      percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                 57.1                   0.0        0.0          63915.0   
1                  0.0                   0.0        0.0         136425.0   
2                  0.0                   0.0        0.0          82124.0   
3                 20.0                   0.0        0.0          54855.0   
4                 75.0                   0.0        0.0          90445.0   
...                ...                   ...        ...              ...   
4697              28.6                   0.0        0.0         159688.0   
4698               0.0                   0.0        0.0          62375.0   
4699               0.0                   0.0        0.0          43250.0   
4700               0.0                   1.0        0.0          31357.0   
4701              33.3                   0.0        0.0         218686.0   

     total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0              49510.0         49400.0                     14515.0   
1              19439.0         15500.0                     18925.0   
2              65000.0          5400.0                     61724.0   
3              50335.0         23200.0                     26255.0   
4              56541.0         15300.0                     72345.0   
...                ...             ...                         ...   
4697          110873.0         48400.0                    107388.0   
4698           18928.0         13300.0                     30775.0   
4699           33022.0          8500.0                     29550.0   
4700           19595.0          1500.0                      9657.0   
4701          209389.0          5600.0                    209986.0   

      hardship_flag  debt_settlement_flag     target  
0                 Y                     N   low_risk  
1                 N                     N   low_risk  
2                 Y                     N   low_risk  
3                 N                     N   low_risk  
4                 N                     N   low_risk  
...             ...                   ...        ...  
4697              N                     N  high_risk  
4698              N   

In [104]:
# Drop the target column and create a features DataFrame
X_q1 = q1_loans_df.drop('target', axis=1) 
X_q1.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    40000.0    0.1033       856.40           RENT    128700.0   
1    24450.0    0.1430       572.72       MORTGAGE     44574.0   
2    13500.0    0.1430       316.23            OWN     60000.0   
3    10625.0    0.1774       268.31           RENT     60000.0   
4     6375.0    0.1862       232.46           RENT     60000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0     Source Verified          n  12.47          0.0             1.0  ...   
1        Not Verified          n  15.05          0.0             1.0  ...   
2        Not Verified          n  28.72          0.0             0.0  ...   
3            Verified          n  15.70          0.0             4.0  ...   
4     Source Verified          n  35.50          0.0             0.0  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            84.2              57.1                   0.0        0.0   
1           100.0               0.0                   0.0        0.0   
2           100.0               0.0                   0.0        0.0   
3            91.4              20.0                   0.0        0.0   
4           100.0              75.0                   0.0        0.0   

  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0         63915.0            49510.0         49400.0   
1        136425.0            19439.0         15500.0   
2         82124.0            65000.0          5400.0   
3         54855.0            50335.0         23200.0   
4         90445.0            56541.0         15300.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     14515.0              Y                     N  
1                     18925.0              N                     N  
2                     61724.0              Y                     N  
3                     26255.0              N                     N  
4                     72345.0              N                     N  

[5 rows x 83 columns]

In [106]:
# Use One-hot encoding for the entire dataframe for 2019 loan data
X_py_dummies = pd.get_dummies(X_py)
print (X_py_dummies.columns)
X_py_dummies.head()

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   
2    11000.0    0.2055       294.81     45000.0  37.25          1.0   
3     4000.0    0.1612       140.87     38000.0  42.89          1.0   
4    14000.0    0.1797       505.93     43000.0  22.16          1.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   
2             3.0      23.0      0.0     8242.0  ...   
3             0.0       7.0      0.0    12767.0  ...   
4             0.0      22.0      0.0    11182.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   
2                             1             1                      0   
3                             0             1                      0   
4                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   
2                      1                            1   
3                      1                            0   
4                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   
2                           0                1                0   
3                           1                1                0   
4                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  
2                       1                       0  
3                       1                       0  
4                       1                       0  

[5 rows x 92 columns]

In [107]:
# Use One-hot encoding for the entire dataframe for 2020 Q1 loan data
X_q1_dummies = pd.get_dummies(X_q1)
print (X_q1_dummies.columns)
X_q1_dummies.head()

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    40000.0    0.1033       856.40    128700.0  12.47          0.0   
1    24450.0    0.1430       572.72     44574.0  15.05          0.0   
2    13500.0    0.1430       316.23     60000.0  28.72          0.0   
3    10625.0    0.1774       268.31     60000.0  15.70          0.0   
4     6375.0    0.1862       232.46     60000.0  35.50          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             1.0       8.0      0.0    38113.0  ...   
1             1.0       6.0      0.0     1665.0  ...   
2             0.0       8.0      0.0    13857.0  ...   
3             4.0      17.0      0.0     6216.0  ...   
4             0.0      13.0      0.0    12681.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    1                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             1   
4                                    1                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                0   
1                            1                           0                1   
2                            1                           0                0   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                1                       1  
1                0                       1  
2                1                       1  
3                0                       1  
4                0                       1  

[5 rows x 91 columns]

In [108]:
# In 2020 Q1 data, there is a column missing.  debt_settlement_flag_Y.  
X_py_dummies['debt_settlement_flag_Y'].value_counts()

0    12175
1        5
Name: debt_settlement_flag_Y, dtype: int64

In [109]:
# This column needs to be added to make sure the model runs correctly
# Add the column and set the values to 0 which reflect a negative response in the field.
X_q1_dummies['debt_settlement_flag_Y']=0
X_q1_dummies.columns

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

In [110]:
# create the target DataFrames for both 2019 and 2020 Q1 data
y_py = py_loans_df['target']
y_q1 = q1_loans_df['target']

In [111]:
# Split the 2019 data into train and test
X_py_train, X_py_test, y_py_train, y_py_test = train_test_split(X_py_dummies, y_py, random_state=1)

In [112]:
# Create a lotistic regression model
classifier = LogisticRegression()
classifier

LogisticRegression()

In [113]:
# Fit (train) model by using the 2019 training data
classifier.fit(X_py_train, y_py_train)

C:\Users\jcbat\anaconda\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [116]:
# print the 2019 training score
print(f"Training Data Score for 2019 train using logistic regression: {classifier.score(X_py_train, y_py_train)}")

Training Data Score for 2019 train using logistic regression: 0.6513409961685823


In [115]:
# Split the 2020 Q1 data into train and test
X_q1_train, X_q1_test, y_q1_train, y_q1_test = train_test_split(X_q1_dummies, y_q1, random_state=1)

In [117]:
# Print the 2020 Q1 test score
print(f"Testing Data Score for 2020 Q1 test using logistic regression: {classifier.score(X_q1_test, y_q1_test)}")

Testing Data Score for 2020 Q1 test using logistic regression: 0.5289115646258503


In [118]:
# Create a random forest model, fitting and training using the 2019 training data
clf = RandomForestClassifier(random_state=1, n_estimators=800).fit(X_py_train, y_py_train)

In [119]:
# Print the 2019 train score
print(f"The Random Forest training score for 2019 data: {clf.score(X_py_train, y_py_train)}")

The Random Forest training score for 2019 data: 1.0


In [120]:
# Print the 2020 Q1 test score
print(f"The Random Forest test score for 2020 Q1 data: {clf.score(X_q1_test, y_q1_test)}")

The Random Forest test score for 2020 Q1 data: 0.6666666666666666


### Conclusion Phase 1

Logistic Regression
* Train Score = 0.65
* Test Score = 0.53

Random Forest
* Train Score = 1.0
* Test Score = 0.67

The hypothesis was proven to be correct in this example.  The random forest classifier provided a better outcome on the 2020 Q1 test data than the logistic regression.  

## Phase 2: Logistic Regression vs. Random Forest Hypothesis (scaled data)

Hypothesis - the random forest classifier will still be the best metho even with the scaled data due to the same factors.  
The scaling will do little to improve the scores.  

In [122]:
# Scale the data
scaler = StandardScaler().fit(X_py_train)
X_py_train_scaled = scaler.transform(X_py_train)
X_q1_test_scaled = scaler.transform(X_q1_test)

In [123]:
# Print the 2019 train score with logistic regression using scaled data
print(f"2019 train score logistic regression with scaled data: {classifier.score(X_py_train_scaled, y_py_train)}")

2019 train score logistic regression with scaled data: 0.6317460317460317


C:\Users\jcbat\anaconda\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [124]:
# Print the 2020 Q1 test score with logistic regression using scaled data
print(f"2020 Q1 test score logistic regression with scaled data: {classifier.score(X_q1_test_scaled, y_q1_test)}")

2020 Q1 test score logistic regression with scaled data: 0.5102040816326531


C:\Users\jcbat\anaconda\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [127]:
# Create a random forest model, fitting and training using the 2019 scaled training data
clf_scaled = RandomForestClassifier(random_state=1, n_estimators=800).fit(X_py_train_scaled, y_py_train)

In [128]:
# Print the 2019 training score with random forest using scaled data
print(f"2019 training score random forest with scaled data: {clf_scaled.score(X_py_train_scaled, y_py_train)}")

2019 training score random forest with scaled data: 1.0


In [129]:
# Print the 2020 Q1 test score with random forest using scaled data
print(f"2020 Q1 test score random forest with scaled data: {clf_scaled.score(X_q1_test_scaled, y_q1_test)}")

2020 Q1 test score random forest with scaled data: 0.6666666666666666


### Conclusion Phase 2

Logistic Regression - Scaled
* Train Score = 0.63
* Test Score = 0.51

Random Forest
* Train Score = 1.0
* Test Score = 0.67

The hypothesis was proven to be correct in this example.  The random forest classifier had a better score than the logistic regression.

## Final Conclusions

In both the unscaled and scaled data, the random forest classifier performed slightly better than the logistic regression model.  Further analysis could determine what features are significant to the target, and if those features do or do not have a linear relationship to the target.  These results drive an opinion that there is not a strong linear relationship between the key features and the target.  

Going from unscaled data to scaled data did not have a significant impact on the score of the logistic regression model (only 0.02 point improvement with scaled data).  The scaling did not change the fundamental nature of the data, therefore did not make the data more linear for the modeling.  

The random forest model did not have any change with scaled data versus unscaled.  The decisions trees from the data did not alter enough to drive a change in the scoring with the data.  That would make sense - as the underlying choices in the decision tree modeling did not change with the scaled data.